lightning/src/ln/channelmanager.rs

```diff
@@ -433,17 +435,52 @@ pub struct ChannelManager<ChanSigner: ChannelKeys, M: Deref, T: Deref, K: Deref,
 	/// new channel.
 	per_peer_state: RwLock<HashMap<PublicKey, Mutex<PeerState>>>,
 
+	persistence_notifier: PersistenceNotifier,
+
 	pending_events: Mutex<Vec<events::Event>>,
 	/// Used when we have to take a BIG lock to make sure everything is self-consistent.
 	/// Essentially just when we're serializing ourselves out.
 	/// Taken first everywhere where we are making changes before any other locks.
+	/// When acquiring this lock in read mode, rather than acquiring it directly, call
+	/// `ReadLockNotifierGuard::new(..)` and pass the lock to it, to ensure its desired
+	/// callback gets executed when the lock is released.
 	total_consistency_lock: RwLock<()>,
 
 	keys_manager: K,
 
 	logger: L,
 }
 
+/// Whenever we release the `ChannelManager`'s `total_consistency_lock`, from read mode, it is
+/// desirable to notify any listeners on `wait_timeout`/`wait` that new updates are available for
+/// persistence. Therefore, this struct is responsible for locking the total consistency lock and,
+/// upon going out of scope, sending the aforementioned notification (since the lock being released
+/// indicates that the updates are ready for persistence).
+struct ReadLockNotifierGuard<'a> {
```

Comment:
I'm having some second thoughts on the naming. My trepidation stems from the fact that this lock is not really used for reading in the traditional sense. It's really saying, "I'm modifying state. So don't allow anyone to persist unless nobody is modifying state." So in that sense locking is really an implementation detail that doesn't need to be exposed in the guard name. Instead, I would propose naming this `PersistenceNotifierGuard`.

However, taking a step back, I'm wondering if `total_consistency_lock` and `persistence_notifier` should be combined as `RwLock<PersistenceNotifier>` instead. Currently, `total_consistency_lock` is being used to allow either shared access for notification or exclusive access for the purpose of persisting. So then does it make sense to combine these in this or some other manner?

I imagine we'd still have a guard struct, perhaps named `PersistenceNotificationGuard` to signify `notify` is called when the guard is dropped.

Reply:
> I'm wondering if total_consistency_lock and persistence_notifier should be combined as RwLock<PersistenceNotifier> instead

I agree, they seem pretty entwined so I added this. Also renamed to `PersistenceNotifierGuard`.